src/utils.rs

```diff
@@ -16,37 +41,287 @@ pub fn generate_eq<F: ExtensionField<F>>(points: &Vec<F>) -> Vec<F> {
     }
 
     res
 }
 
 // Algorithm 5
-pub fn initialize_phase_two<F: Field, E: ExtensionField<F>>(
-    igz: Vec<E>,
-    iux: Vec<E>,
-    f1: Vec<(usize, usize, usize)>,
+pub(crate) fn initialize_phase_two<F: Field, E: ExtensionField<F>>(
+    igz: &[E],
+    iux: &[E],
+    f1: &[(usize, usize, usize)],
 ) -> Vec<E> {
     let mut res = vec![E::zero(); iux.len()];
 
     for (z, x, y) in f1 {
         // It is assumed the operation poly outputs 1 where there is a valid gate
-        res[y] += igz[z] * iux[x];
+        res[*y] += igz[*z] * iux[*x];
     }
 
     res
 }
 
+// Combines a vector of sumcheck proofs into one sumcheck proof
+pub fn combine_sumcheck_proofs<F: Field, E: ExtensionField<F>>(
+    sumcheck_proofs: Vec<SumCheckProof<F, E>>,
+) -> SumCheckProof<F, E> {
+    sumcheck_proofs
+        .into_iter()
+        .reduce(|mut acc, val| {
+            acc.round_polynomials
+                .extend_from_slice(&val.round_polynomials);
+            acc
+        })
+        .unwrap()
+}
+
+pub(crate) fn prepare_phase_one_params<F: Field, E: ExtensionField<F>>(
+    g: &[E],
+    add_i: &[(usize, usize, usize)],
+    mul_i: &[(usize, usize, usize)],
+    w_i_plus_one: &[F],
+) -> (Vec<E>, Vec<E>, Vec<E>, Vec<E>) {
+    let igz = generate_eq(g);
+
+    let ident = vec![F::one(); w_i_plus_one.len()];
+
+    // Build Ahg for mul, add_b and add_c
+    let mul_ahg = initialize_phase_one(&igz, mul_i, w_i_plus_one);
+
+    let add_b_ahg = initialize_phase_one(&igz, add_i, &ident);
+
+    let add_c_ahg = initialize_phase_one(&igz, add_i, w_i_plus_one);
+
+    (igz, mul_ahg, add_b_ahg, add_c_ahg)
+}
+
+// hg(x)
+pub(crate) fn build_phase_one_libra_sumcheck_poly<F: Field, E: ExtensionField<F>>(
+    mul_ahg: &[E],
+    add_b_ahg: &[E],
+    add_c_ahg: &[E],
+    w_i_plus_one_poly: &MultilinearPoly<F, E>,
+) -> VPoly<F, E> {
+    let n_vars = w_i_plus_one_poly.num_vars();
+
+    VPoly::new(
+        vec![
+            MultilinearPoly::new_from_vec(
+                n_vars,
+                mul_ahg
+                    .iter()
+                    .map(|val| Fields::<F, E>::Extension(*val))
+                    .collect::<Vec<Fields<F, E>>>(),
+            ),
+            MultilinearPoly::new_from_vec(
+                n_vars,
+                add_b_ahg
+                    .iter()
+                    .map(|val| Fields::<F, E>::Extension(*val))
+                    .collect::<Vec<Fields<F, E>>>(),
+            ),
+            MultilinearPoly::new_from_vec(
+                n_vars,
+                add_c_ahg
+                    .iter()
+                    .map(|val| Fields::<F, E>::Extension(*val))
+                    .collect::<Vec<Fields<F, E>>>(),
+            ),
+            w_i_plus_one_poly.clone(),
+        ],
+        2,
+        n_vars,
+        Rc::new(|data: &[Fields<F, E>]| (data[3] * (data[0] + data[1])) + data[2]),
+    )
+}
+
+pub(crate) fn prepare_phase_one_params_with_alpha_beta_rb_rc<F: Field, E: ExtensionField<F>>(
```

Comment:
there is some duplication between this function and `prepare_phase_one_params` seeing the potential for a sub function that takes in Igz. 
so prepare_phase_one_params builds Igz directly
and this builds igz with alpha beta
but both make use of the same sub function